### ISSUE TITLE
Implement UI v3 with React

### ISSUE THREAD

**Issue by gimite** [google/personfinder#478]
@nworden I created this because I couldn't find an issue for the overall UI refresh effort. But let me know if it is already somewhere.

I arbitrarily called it "v3" because the current UI was called "v2". Feel free to give it your preferred name :)

**Comment by yashLadha** [comment]
Hey I would like to work on this issue and work alongside with @nworden . @gimite

**Comment by nworden** [comment]
Hi Yash,
I'm sorry for my delay in replying (I was offline yesterday). Thanks for your offer to help!
Are you already familiar with React? If not, the tutorial on their website is pretty good so I'd suggest starting there. If you're comfortable with React already, let me know and we can find a piece of the UI work to break off for you.
- Nick

**Comment by yashLadha** [comment]
@nworden  Yes, I am comfortable with React and had created a couple of personal projects in react+redux as well. So I would be happy to contribute to this project.

**Comment by nworden** [comment]
Ok, that's great. Here are a few options that might be good places to start:
- A component for our static pages. We have a few pages that display static information (e.g., [the page for responders](https://google.org/personfinder/global/responders.html)), and we'll need to support them in the React UI. The implementation for each of these pages will probably be very similar, so I think the way to do it would be to create a general StaticPage component and subclass it for specific pages. [Here's a mock of one of the pages.](https://user-images.githubusercontent.com/9059304/52650900-d19a6d00-2ef3-11e9-9fad-8e742efbd590.png)
- Support a turndown message. When a repository for an event is getting less traffic and gets scheduled to be turned down, we want to announce the upcoming turndown to users. This would involve some backend work too, to pass the fact of the impending deactivation to the frontend.

Does one of those sound like something you'd be interested in doing? Don't hesitate to let me know if you have any questions!

**Comment by yashLadha** [comment]
Sounds great. I will start working on the general static component ASAP and can you provide me with some more insight on the second task, as it will be a good task for learning purpose. @nworden . I am creating a separate issue for the first point. Also how about using typescript instead of javascript for frontend in React. This can possibly avoid a lot of bugs in production.

**Comment by nworden** [comment]
Regarding your question on PR #526 (just answering here to try to keep it in one place): I don't think I see a need for a HOC here. IMO the way to do it would be to create a run-of-the-mill component that uses a prop for the title and `children` for the actual page contents (for now, the contents are just going to be a react-intl message, but doing it via `children` would be useful if we ever want to included other components within them in the futuure). Then you could either include that component directly in App.js (with the appropriate title and message for the page) or create light wrapper-y components around it for each page. What do you think?

Regarding the second task: first, for clarity, the message feature isn't really specific to turndown messages (and I should really stop referring to it that way); it just happens that turndown messages are the main thing they're used for. What it is is a feature where administrators can specify custom messages to display in certain places (e.g., on the repo homepage or a search results page). If you go to <your PF instance's address>/<repo name>/admin, you'll see an admin page. One section is labeled "Optional custom messages", with fields for all the places custom messages can appear, for each language used for the repo. Those messages will need to get passed through to the React UI (from the /d/repo call, which will be implemented [here](https://github.com/google/personfinder/blob/master/app/frontend_api.py#L22)) and displayed in the UI for their repo. Right now /d/repo isn't really implemented at all yet, so I would suggest just hard-coding dummy return data (similar to what's already there) if you want to start the frontend work right away, and hopefully I'll have had a chance to start the real /d/repo implementation before you get to the backend work.

Lastly, regarding TypeScript: TBH I hadn't considered using TS, but I don't think we should. The main reason for this is that Person Finder's a small project with a small team, and it's preferable to stick to more widely used technologies unless we have a strong reason not to, so that it's easier to ensure that we have multiple team members who are familiar with each technology we use. (Although it wasn't the main reason, this is also part of why we went with React over Angular.) Also, it's honestly not a really complex app, and I haven't encountered any typing-related bugs (even during development). If we find ourselves building something more complex, it'd be worth reconsidering (and I might regret my choice when I have to migrate it), but I don't think TS is worth the added complexity for us.

**Comment by yashLadha** [comment]
Thanks for the detailed explanation @nworden . I will send a PR for a static component ASAP.

**Comment by yashLadha** [comment]
@nworden  Can you just define the props that need to be rendered in the single component like the basic entities. This will help me to formulate the structure for the component and can send PR after that. Currently, I am just using a single component which is rendering the UI from the passed props. Also is the passed props will be HTML or some data.

**Comment by nworden** [comment]
Sure. For now I think the page title (or subtitle, depending on how you look at it) and the page contents are the only thing you really need to deal with. It'd probably be easiest to pass the contents as the component's children rather than as a "regular" prop. For example, here's how we might use the static component for the new version of [this page](https://google.org/personfinder/global/responders.html) (the one that's mocked in the image a few comments up):

```javascript
<YourGeneralStaticComponent
    subtitle="Help document">
  <h2>Information for responders</h2>
  <p>Any time Google launches... etc.</p>
  <h3>Embed Google Person Finder in your site</h3>
  <p>Spread the word about... etc.</p>
  and so on
</YourGeneralStaticComponent>
```

Is that helpful?

The static component would be responsible for rendering the header section (with the specified subtitle) and the footer (`components/Footer.js`), with the children in between them.

Something I just realized is that the static pages have a header that looks basically the same as the header for repository pages (for which I implemented `components/RepoHeader.js`). It'd probably be a good idea to re-use `RepoHeader` for the static pages too: we could rename it to something more general like `PageHeader`, and change it to accept `title` directly as a prop, and then it could be used for headers of non-repository pages too. I can make the changes to `RepoHeader` if you like, or you can do it if you prefer, just let me know.

**Comment by sumanthratna** [comment]
Just curious, but why wasn't Angular chosen? This doesn't really count as a reason, but I feel like Google typically uses Google-developed tools. 

On a more serious note, I personally feel like the organization of the components in this project would make more sense with Angular 2 and Material.

**Comment by nworden** [comment]
> Just curious, but why wasn't Angular chosen? This doesn't really count as a reason, but I feel like Google typically uses Google-developed tools.

We considered Angular, but there's a couple reasons we chose React instead, mainly:
- Angular's more heavyweight, which I'm sure is great for some projects, but it was just more than we needed (we're perhaps more averse to using extra bandwidth than some Google projects are, since PF's often used in situations with limited connectivity, shared wi-fi, etc.).
- We're open-source, and (outside Google) React is used more widely than Angular is.
- The options for using Angular to implement a Material UI were not what one might expect: the Material team doesn't provide open-source Angular components, but does provide React components. For Angular, they recommend a third-party library that's almost entirely maintained by one individual, which is something we try to stay away from (we prefer dependencies with some kind of organization/community behind them, especially if it's something that'd be hard to switch out for something else).

> On a more serious note, I personally feel like the organization of the components in this project would make more sense with Angular 2 and Material.

I'm not sure exactly what you mean here?
